### DISCUSSION TITLE
Dotted barlines

### DISCUSSION THREAD

**Discussion by craigsapp** [rism-digital/verovio#2839] [Ideas]
Dotted barlines are not dotted, but rather tiny dashes:
<img width="850" alt="Screen Shot 2022-05-01 at 8 14 25 PM" src="https://user-images.githubusercontent.com/3487289/166180023-79a5a46a-9df3-4713-80f2-60a4cbe2d1ad.png">

<details>
<summary>
Click to view MEI data for above example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-01T19:59:10" version="3.10.0-dev-141ef9e">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mz2febq">
    <score xml:id="sshp054">
     <scoreDef xml:id="s6n7qz2" midi.bpm="400.000000">
      <staffGrp xml:id="sl9lid9" bar.thru="true" symbol="brace">
       <staffDef xml:id="sgj102o" n="1" lines="5">
        <clef xml:id="clef-L3F2" shape="G" line="2" />
        <meterSig xml:id="metersig-L2F2" count="4" unit="4" />
       </staffDef>
       <staffDef xml:id="sl9ordu" n="2" lines="5">
        <clef xml:id="clef-L3F1" shape="F" line="4" />
        <meterSig xml:id="metersig-L2F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" right="dashed" n="1">
       <staff xml:id="sdxkc04" n="1">
        <layer xml:id="layer-L1F2N1" n="1">
         <note xml:id="note-L5F2" dur="1" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="sq0m2ba" n="2">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="1" oct="3" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L6" right="dbldashed" n="2">
       <staff xml:id="staff-L6F2N1" n="1">
        <layer xml:id="layer-L6F2N1" n="1">
         <note xml:id="note-L7F2" dur="1" oct="4" pname="b" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-L6F1N1" n="2">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L7F1" dur="1" oct="3" pname="d" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L8" right="dotted" n="3">
       <staff xml:id="staff-L8F2N1" n="1">
        <layer xml:id="layer-L8F2N1" n="1">
         <note xml:id="note-L9F2" dur="1" oct="4" pname="a" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-L8F1N1" n="2">
        <layer xml:id="layer-L8F1N1" n="1">
         <note xml:id="note-L9F1" dur="1" oct="3" pname="e" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L10" right="dbldotted" n="3">
       <staff xml:id="staff-L10F2N1" n="1">
        <layer xml:id="layer-L10F2N1" n="1">
         <note xml:id="note-L11F2" dur="1" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-L10F1N1" n="2">
        <layer xml:id="layer-L10F1N1" n="1">
         <note xml:id="note-L11F1" dur="1" oct="3" pname="f" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L12" right="end" n="3">
       <staff xml:id="staff-L12F2N1" n="1">
        <layer xml:id="layer-L12F2N1" n="1">
         <note xml:id="note-L13F2" dur="1" oct="4" pname="f" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-L12F1N1" n="2">
        <layer xml:id="layer-L12F1N1" n="1">
         <note xml:id="note-L13F1" dur="1" oct="3" pname="g" accid.ges="n" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

At a small scale they look like plain barlines.   Here is a zoom into the dotted line:

<img width="208" alt="Screen Shot 2022-05-01 at 8 16 54 PM" src="https://user-images.githubusercontent.com/3487289/166180169-1f1bf343-963f-4d83-b6fb-368820dee3f8.png">

Those do not look like dots :-)

Ideally dotted lines would look something like this:

<img width="1039" alt="Screen Shot 2022-05-01 at 8 15 13 PM" src="https://user-images.githubusercontent.com/3487289/166180078-eace7c65-9975-4000-b818-d2b1b80679cb.png">

**Comment by rettinghaus** [comment]
@craigsapp I don't like the current look of dotted barlines, but I'm not sure your suggestion is a big improvement. 
How do dotted lines look like in SCORE? 
For comparison:
 
### LilyPond

The only _real_ dotted barline I can think of 

![image](https://user-images.githubusercontent.com/7693447/166457724-a56d1fab-5eeb-411c-9ecb-05c30d3d6104.png)

### MuseScore

Pretty much looks like Verovio's approach

![image](https://user-images.githubusercontent.com/7693447/166456352-d3946895-0ffe-47b5-81f5-c60b2401d0fa.png)

### Dorico

Dorico doesn't offer dotted barlines

### Gould

Gould calls dashed barlines "dotted" barlines

![image](https://user-images.githubusercontent.com/7693447/166456551-f365a33b-8ac7-47c5-b411-a8eb043d544a.png)

### Schott

in Stravinsky's _Dumbarton Oaks_ they place dots between the staff lines

![image](https://user-images.githubusercontent.com/7693447/166459158-670b8e2c-89fa-45d9-9103-a77ce74cce71.png)

### Breitkopf und Härtel

in the _Fitzwilliam Virginal Book_ we found these:

![image](https://user-images.githubusercontent.com/7693447/166461491-43239866-aba5-42a2-ab29-ed346e0d4c35.png)

So I'm wondering if it would be good to use repetition dots for dotted barlines?

**Comment by craigsapp** [comment]
>  but I'm not sure your suggestion is a big improvement.

My example is infinitely better than the current verovio dotted lines, because I give a score of 0 to the current verovio dotted lines, and my example gets a non-zero score 😛 

I like the single-dot per space of the Schott and Breitkof & Härtel examples.  The Lilypond dots seem too large (or maybe that is just the effect of the staff being larger than the other two samples).   A complication with this system is that when drawing dotted lines across staves (as in B&H example), the spacing between the staves should be an integer number of staff lines (tiny adjustments could be made to the dot spacings between staves to make them align with the spaces between adjacent staves, however).

SCORE does not have dotted barlines, so in the above example I made (in SCORE), the barline is invisible and I superimpose a dotted line on the barline.   For dotted lines and and dashed barlines, there are parameters to control the spacing/size of the lines.

For dotted lines, you can control the size of the dots and their spacings.  Here are some examples:

<img width="274" alt="Screen Shot 2022-05-03 at 7 40 18 AM" src="https://user-images.githubusercontent.com/3487289/166475321-fa7aecf4-da29-45f1-8e98-43e3b1c67193.png">

I made the first one fit three dots to a staff space, the second has 4 dots per staff space and their diameters are set to 1/4 staff space so that they touch.   The third one is a single dot per space at the default thickness, and the last one is the original examples 2 dots per space at a smaller size.

Here is a demonstration of staff line controls in SCORE:

<img width="334" alt="Screen Shot 2022-05-03 at 7 39 51 AM" src="https://user-images.githubusercontent.com/3487289/166475827-b02e9f5b-681e-42b1-8321-0f08b2837349.png">

The original one is on the right, and then I adjust the dash lengths, and for the last two I adjust the spacing length.  SCORE will always do dashed barlines that have a complete dash at the top and bottom, so it tweaks the actual dash/space sizes to fit.

<details>
<summary>
SCORE PMX data for dashed/dotted barline samples
</summary>

```tsv
8  1   0.000  0.00  0.00 200.00
8  2   0.000  0.00  0.00 200.00
14 1   0.000  2.00
14 1   0.000  2.00  8.00
3  2   2.000
3  1   2.000  0.00  1.00
18 1   9.500  0.00  4.00   4.00
18 2   9.500  0.00  4.00   4.00
1  1  17.000  6.00 10.00   2.00  4
1  2  17.000  8.00 20.00   2.00  4
14 1  32.827  2.00  7.00   0.00  0 6.00 6.000
14 1  35.827  2.00  7.00   0.00  0 6.00 3.000
14 1  38.827  2.00  7.00   0.00  0 6.00
14 1  41.827  2.00  7.00   0.00  0 5.00
14 1  44.827  2.00  7.00   0.00  0 3.00
14 1  47.827  2.00  7.00   0.00  0 2.00
14 1  50.827  2.00  7.00
1  1  54.285  7.00 20.00   2.00  4
1  2  54.285  7.00 20.00   2.00  4
14 1  87.048  2.00  7.00
14 1  88.048  2.00  7.00
1  2  91.505  6.00 10.00   2.00  4
1  1  91.505  8.00 20.00   2.00  4
4  2 115.750 10.75 -0.25 145.75 -4 0.25 0.667 0.1 0 0 90
4  2 119.000 11.00  0.00 149.00 -4 0.00 0.500 0.1 0 0 90
4  2 122.000 10.25 -0.75 152.00 -4 0.00 0.000 0.0 0 0 90
4  2 125.000 10.75 -0.25 155.00 -4 0.50 1.050 0.0 0 0 90
1  2 128.790  5.00 10.00   2.00  4
1  1 128.790  9.00 20.00   2.00  4
4  2 161.000 10.75 -0.25 191.00 -4 0.50 1.050 0.0 0 0 90
4  2 162.250 10.75 -0.25 192.25 -4 0.50 1.050 0.0 0 0 90
1  2 166.075  4.00 10.00   2.00  4
1  1 166.075 10.00 20.00   2.00  4
14 1 200.000  2.00  2.00
```

</details>
